addons/account/models/account_journal.py

```diff
@@ -600,9 +600,19 @@ def _alias_prepare_alias_name(self, alias_name, name, code, jtype, company):
         )
         if company != self.env.ref('base.main_company'):
             company_identifier = company.name if self.env['mail.alias']._is_encodable(company.name) else company.id
-            if f'-{company_identifier}' not in alias_name:
-                alias_name = f"{alias_name}-{company_identifier}"
-        return self.env['mail.alias']._sanitize_alias_name(alias_name)
+            if f'-{company_identifier}' not in base_alias_name:
+                base_alias_name = f'{base_alias_name}-{company_identifier}'
+
+        # Check for duplicate alias, if exists append a suffix to ensure uniqueness
+        # It's not a duplicate if it's the same journal
+        sanitized_alias_name = self.env['mail.alias']._sanitize_alias_name(base_alias_name)
+        existing_alias = self.env['mail.alias'].search([('alias_name', '=', sanitized_alias_name)])
+        if len(existing_alias) > 0 and existing_alias.id != self.alias_id.id:
+            alias_name = f'{sanitized_alias_name}-{code}-{company.id}'
```

Comment:
But code is already taken into account in base_alias_name ?

Reply:
In the failing upgrades (around 17 since last week), when I tested in a fresh DB, and from reading the code (AFAICT) the mail alias generated from the journal doesn't have a code,  journal 'Customer Invoices' would generate 'customer-invoices' as the alias name for example, and the full alias in most of the failing cases is `-------@localhost`

> Please also note that you should provide some unit tests in order to highlight the issue, otherwise it is a bit hard to see the final result :) .

Ok! Indeed. :)

Reply:
"-------@localhost" ? Encoding issue ?

Reply:
Oh, you tagged the user localhost xD
I used "----" to ignore the less important part, i meant to say that the domain is most of the time `localhost`, and when it's not, it's still not unique, this is in the context of same company, the error is easy to reproduce..

But I didn't know about the uniqueness in a given alias domain, I will fix things and add a unit test!

Reply:
> But you talk about upgrade. However alias names should be unique already in 16, did something change ?

They are, but the way they are created isn't always unique.

As said in the commit message, the issue is mainly when the name of a journal is changed, because that doesn't change the name of the alias that was created with it:
- I create a journal named `A`, automatically an alias named `A` will be created. Afterwards if the journal name is changed, the alias will remain `A`. (changing the journal name doesn't affect the alias name)
Now a new journal named `A` can be created because that name is not taken. This will fail because the creation of that journal will trigger the creation of an alias `A` which is already exists.
This also happens when  users changed or manually translated the name of journals that where created by odoo because it doesn't suit their needs. As during the upgrade, Odoo will try to recreate the journals since the name is not the same anymore.

Reply:
I see two different issues then
* create a journal, rename it, create a new journal with old name -> that's effectively a standard issue; would say we probably don't care in standard addons (there is few automatic alias name generation), but we can effectively improve accounting (but still not sure for the heuristic here, to check how accounting see it; shouldn't we add a number as suffix ?)
* upgrade trying to create journals again -> isn't it an upgrade issue ? Should it do it ?

Reply:
> shouldn't we add a number as suffix ?
That was the first version of the code, but the combination of `code` and `company_id` is unique, so we can have a unique alias without having numbering like 1,2,3...

> upgrade trying to create journals again -> isn't it an upgrade issue ? Should it do it ?
During the upgrade, as in the creation of a new odoo DB, data is added from data/demo directories to the DB, that's normal

The issue with the account journals made by the chart template is that they don't have an xmlid, see here how they are created:
https://github.com/odoo/odoo/blob/17.0/addons/account/models/chart_template.py#L913
(Currently looking into why they are made this way)

So almost any change to these template journals will make them `unrecognisable`, when the chart is loaded again or during an upgrade, it will seem like data created by the user.